src/object/task/taskgoto.cpp

```diff
@@ -1698,199 +1727,326 @@ void CTaskGoto::BeamInit()
+            else
+            {
+                if (totalDistance < m_bfsQueueMin + NUMQUEUEBUCKETS)
+                {
+                    // Move node to a regular bucket.
+                    m_bfsQueue[totalDistance % NUMQUEUEBUCKETS].push_back(indexInMap);
+                    m_bfsQueueCountPushed += 1;
+                    GetLogger()->Debug("Moving node with bigger distance into regular bucket, distance: %d, totalDistance: %d, m_bfsQueueMin: %d\n",
+                        distance, totalDistance, m_bfsQueueMin);
+                }
+                else
+                {
+                    // Move node to the bucket with oversized costs.
+                    m_bfsQueue[NUMQUEUEBUCKETS].push_back(indexInMap);
+                    m_bfsQueueCountPushed += 1;
+                    GetLogger()->Debug("Moving node with bigger distance into bucket with oversized costs, distance: %d, totalDistance: %d, m_bfsQueueMin: %d\n",
+                        distance, totalDistance, m_bfsQueueMin);
+                }
+            }
+            continue;
         }
-        iLar ++;
-    }
 
-    return ERR_GOTO_IMPOSSIBLE;
-}
+        if (x == startX && y == startY)
+        {
+            // We have reached the start.
+            // Follow decreasing distances to find the path.
+            m_bmPoints[0] = start;
+            int btX = x;
+            int btY = y;
+            for (m_bmTotal = 1; m_bmTotal < MAXPOINTS; ++m_bmTotal)
+            {
+                int bestX = -1;
+                int bestY = -1;
+                int32_t bestDistance = std::numeric_limits<int32_t>::max();
+                for (int i = 0; i < 8; ++i)
+                {
+                    const int nX = btX + dXs[i];
+                    const int nY = btY + dYs[i];
+                    if (!BitmapTestDot(1, nX, nY)) continue;
+                    const int32_t nDistance = m_bfsDistances[nY * m_bmSize + nX];
+                    if (nDistance < bestDistance)
+                    {
+                        bestX = nX;
+                        bestY = nY;
+                        bestDistance = nDistance;
+                    }
+                }
+                if (bestX == -1)
+                {
+                    GetLogger()->Debug("Failed to find node parent\n");
+                    return ERR_GOTO_ITER;
+                }
+                btX = bestX;
+                btY = bestY;
+                if (btX == goalX && btY == goalY)
+                {
+                    m_bmPoints[m_bmTotal] = goal;
+                    break;
+                }
+                m_bmPoints[m_bmTotal].x = (btX + 0.5f) * BM_DIM_STEP - 1600.f;
+                m_bmPoints[m_bmTotal].z = (btY + 0.5f) * BM_DIM_STEP - 1600.f;
 
-// Is a right "start-goal". Calculates the point located at the distance "step"
-// from the point "start" and an angle "angle" with the right.
+                if (bestDistance == 0)
+                {
+                    break;
+                }
```

Comment:
This will set an exact goal position if the `goalRadius` is zero but a rounded position when it isn't. I think this can cause problem when trying to pick up objects. It should be improved before merging.

Reply:
When goalRadius is not 0.0f the last point (m_bmPoints[m_bmTotal]) should be that distance from the position given by parameter 'goal'. This is currently expected during the final phases of the movement part of goto.

Reply:
Yes, but only approximately because it is aligned to the grid. I think the grid cells are big enough to cause issues. The `goalRadius` is 8 for picking up objects and the cell size is 5. Too close for comfort. It would make sense to adjust the distance to the goal in a similar way as it was done before with beam search.

Reply:
Yes, that basically what I was trying to say.